lib/stpeprintf.h

```diff
@@ -0,0 +1,115 @@
+/*
+ * SPDX-FileCopyrightText:  2022 - 2023, Alejandro Colomar <alx@kernel.org>
+ *
+ * SPDX-License-Identifier:  BSD-3-Clause
+ */
+
+
+#ifndef SHADOW_INCLUDE_LIB_STPEPRINTF_H_
+#define SHADOW_INCLUDE_LIB_STPEPRINTF_H_
+
+
+#include <config.h>
+
+#include <stdarg.h>
+#include <stddef.h>
+#include <stdio.h>
+
+
+__attribute__((format(printf, 3, 4)))
+inline char *shdw_stpeprintf(char *dst, char *end, const char *restrict fmt,
+    ...);
+
+__attribute__((format(printf, 3, 0)))
+inline char *shdw_vstpeprintf(char *dst, char *end, const char *restrict fmt,
+    va_list ap);
+
+
+/*
+ * SYNOPSIS
+ *	[[gnu::format(printf, 3, 4)]]
+ *	char *_Nullable stpeprintf(char *_Nullable dst, char end[0],
+ *	                           const char *restrict fmt, ...);
+ *
+ *	[[gnu::format(printf, 3, 0)]]
+ *	char *_Nullable vstpeprintf(char *_Nullable dst, char end[0],
+ *	                           const char *restrict fmt, va_list ap);
+ *
+ *
+ * ARGUMENTS
+ *	dst	Destination buffer where to write a string.
+ *
+ *	end	Pointer to one after the last element of the buffer
+ *		pointed to by `dst`.  Usually, it should be calculated
+ *		as `dst + NITEMS(dst)`.
+ *
+ *	fmt	Format string
+ *
+ *	...
+ *	ap	Variadic argument list
+ *
+ * DESCRIPTION
+ *	These functions are very similar to snprintf(3) and vsnprintf(3).
+ *
+ *	The destincation buffer is limited by a pointer to its end --one
+ *	after its last element-- instead of a size.  This allows
+ *	chaining calls to it safely, unlike snprintf(3), which is
+ *	difficult to chain without invoking Undefined Behavior.
+ *
+ * RETURN VALUE
+ *	dst + strlen(dst)
+ *		•  On success, these functions return a pointer to the
+ *		   terminating NUL byte.
+ *
+ *	end
+ *		•  If this call truncated the resulting string.
+ *		•  If `dst == end` (a previous chained call to these
+ *		   functions truncated).
+ *	NULL
+ *		•  If this function failed (see ERRORS).
+ *		•  If `dst == NULL` (a previous chained call to these
+ *		   functions failed).
+ *
+ * ERRORS
+ *	These functions may fail for the same reasons as vsnprintf(3).
+ */
+
+
+inline char *
+shdw_stpeprintf(char *dst, char *end, const char *restrict fmt, ...)
+{
+	char     *p;
+	va_list  ap;
+
+	va_start(ap, fmt);
+	p = shdw_vstpeprintf(dst, end, fmt, ap);
+	va_end(ap);
+
+	return p;
+}
+
+
+inline char *
+shdw_vstpeprintf(char *dst, char *end, const char *restrict fmt, va_list ap)
+{
+	int        len;
+	ptrdiff_t  size;
+
+	if (dst == end)
+		return end;
+	if (dst == NULL)
+		return NULL;
+
+	size = end - dst;
+	len = vsnprintf(dst, size, fmt, ap);
```

Comment:
`vsnprintf()` or `snprintf()`?

Reply:
`vsnprintf(3)`, it's correct.

```
printf(3)                  Library Functions Manual                  printf(3)

NAME
       printf,  fprintf,  dprintf,  sprintf,  snprintf, vprintf, vfprintf, vd‐
       printf, vsprintf, vsnprintf - formatted output conversion

LIBRARY
       Standard C library (libc, -lc)

SYNOPSIS
       #include <stdio.h>

       ...
       int snprintf(char str[restrict .size], size_t size,
                   const char *restrict format, ...);

       ...
       int vsnprintf(char str[restrict .size], size_t size,
                   const char *restrict format, va_list ap);
```

They are essentially the same, except that snprintf(3) accepts a variable number of arguments, and vsnprintf(3) accepts a `va_list` opaque object, which holds the same information in a different form.  vsnprintf(3) is the one you need for implementing these things internally.  I know I didn't explain it very clearly, but it's not easy for me to explain it :)

I suggest reading `stdarg(3)`, and also this: <https://www.reddit.com/r/learnprogramming/comments/y71z4/snprintf_and_vsnprintf/>


Maybe this also helps.  This is how musl libc implements snprintf(3):

```c
alx@asus5775:~/src/musl/musl$ grepc -tfd snprintf
./src/stdio/snprintf.c:4:
int snprintf(char *restrict s, size_t n, const char *restrict fmt, ...)
{
	int ret;
	va_list ap;
	va_start(ap, fmt);
	ret = vsnprintf(s, n, fmt, ap);
	va_end(ap);
	return ret;
}
```

Reply:
I was asking because the https://github.com/shadow-maint/shadow/pull/636/commits/bd769bd0874e136f6dfe2a4a26b4fee97ee6162b commit message states `snprintf()`, which is kind of confusing.

Reply:
Fixed :)